File: extensions/shared/library/src/main/java/app/revanced/extension/shared/spoof/SpoofSignaturePatch.java
package app.revanced.extension.shared.spoof;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Base64;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Map;

@SuppressWarnings("unused")
public class SpoofSignaturePatch {
    public static void killPM(String packageName, String signatureData) {
        Signature fakeSignature = new Signature(Base64.decode(signatureData, Base64.DEFAULT));
        Parcelable.Creator<PackageInfo> originalCreator = PackageInfo.CREATOR;
        Parcelable.Creator<PackageInfo> creator = new Parcelable.Creator<PackageInfo>() {
            @Override
            public PackageInfo createFromParcel(Parcel source) {
                PackageInfo packageInfo = originalCreator.createFromParcel(source);
                if (packageInfo.packageName.equals(packageName)) {
                    if (packageInfo.signatures != null && packageInfo.signatures.length > 0) {
                        packageInfo.signatures[0] = fakeSignature;
                    }
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                        if (packageInfo.signingInfo != null) {
                            Signature[] signaturesArray = packageInfo.signingInfo.getApkContentsSigners();
                            if (signaturesArray != null && signaturesArray.length > 0) {
                                signaturesArray[0] = fakeSignature;
                            }
                        }
                    }
                }
                return packageInfo;
            }

            @Override
            public PackageInfo[] newArray(int size) {
                return originalCreator.newArray(size);
            }
        };
        try {
            findField(PackageInfo.class, "CREATOR").set(null, creator);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            addHiddenApiExemptions();
        }
        try {
            Object cache = findField(PackageManager.class, "sPackageInfoCache").get(null);
            //noinspection ConstantConditions
            cache.getClass().getMethod("clear").invoke(cache);
        } catch (Throwable ignored) {
        }
        try {
            Map<?, ?> mCreators = (Map<?, ?>) findField(Parcel.class, "mCreators").get(null);
            //noinspection ConstantConditions
            mCreators.clear();
        } catch (Throwable ignored) {
        }
        try {
            Map<?, ?> sPairedCreators = (Map<?, ?>) findField(Parcel.class, "sPairedCreators").get(null);
            //noinspection ConstantConditions
            sPairedCreators.clear();
        } catch (Throwable ignored) {
        }
    }

    private static Field findField(Class<?> clazz, String fieldName) throws NoSuchFieldException {
        try {
            Field field = clazz.getDeclaredField(fieldName);
            field.setAccessible(true);
            return field;
        } catch (NoSuchFieldException e) {
            while (true) {
                clazz = clazz.getSuperclass();
                if (clazz == null || clazz.equals(Object.class)) {
                    break;
                }
                try {
                    Field field = clazz.getDeclaredField(fieldName);
                    field.setAccessible(true);
                    return field;
                } catch (NoSuchFieldException ignored) {
                }
            }
            throw e;
        }
    }

    public static void addHiddenApiExemptions() {
        try {
            Method forName = Class.class.getDeclaredMethod("forName", String.class);
            Method getDeclaredMethod = Class.class.getDeclaredMethod(
                    "getDeclaredMethod",
                    String.class,
                    Class[].class
            );

            Class<?> vmRuntimeClass = (Class<?>) forName.invoke(null, "dalvik.system.VMRuntime");

            Method getRuntime = (Method) getDeclaredMethod.invoke(
                    vmRuntimeClass,
                    "getRuntime",
                    (Object) null
            );

            Method setHiddenApiExemptions = (Method) getDeclaredMethod.invoke(
                    vmRuntimeClass,
                    "setHiddenApiExemptions",
                    new Class[]{String[].class}
            );

            Object vmRuntime = getRuntime.invoke(null);

            setHiddenApiExemptions.invoke(vmRuntime, (Object) new String[]{"L"});
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

}

File: extensions/shared/library/src/main/java/app/revanced/extension/shared/spoof/SpoofSignaturePatch.kt
package app.revanced.extension.shared.spoof

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.os.Build
import android.os.Parcel
import android.os.Parcelable
import android.util.Base64
import java.lang.reflect.Field
import java.lang.reflect.Method

@Suppress("unused")
class SignaturePatch {
    companion object {
        @JvmStatic
        fun killPM(packageName: String, rawSignatureBase64: String) {
            val originalSignature = Signature(Base64.decode(rawSignatureBase64, Base64.DEFAULT))
            val originalCreator = PackageInfo.CREATOR

            val creator = object : Parcelable.Creator<PackageInfo> {
                override fun createFromParcel(source: Parcel?): PackageInfo {
                    var packageInfo = originalCreator.createFromParcel(source)
                    if (packageInfo.packageName != null) {
                        if (packageInfo.packageName.equals(packageName, true)) {
                            packageInfo = spoofSignature(packageInfo, originalSignature)
                        }
                    }
                    return packageInfo
                }

                override fun newArray(size: Int): Array<PackageInfo> {
                    return originalCreator.newArray(size)
                }
            }

            try {
                findField(PackageInfo::class.java, "CREATOR").set(null, creator)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                addHiddenApiExemptions()
            }
            try {
                val cache = findField(PackageManager::class.java, "sPackageInfoCache").get(null)
                cache.javaClass.getDeclaredMethod("clear").invoke(cache)
            } catch (ignored: Throwable) {
            }
            try {
                val mCreators: HashMap<*, *> = findField(Parcel::class.java, "mCreators").get(null) as HashMap<*, *>
                mCreators.clear()
            } catch (ignored: Throwable) {
            }
            try {
                val sPairedCreators: HashMap<*, *> =
                    findField(Parcel::class.java, "sPairedCreators").get(null) as HashMap<*, *>
                sPairedCreators.clear()
            } catch (ignored: Throwable) {
            }
        }

        fun spoofSignature(_packageInfo: PackageInfo, _signature: Signature): PackageInfo {
            var packageInfo = _packageInfo
            if (packageInfo.signatures != null && packageInfo.signatures.size > 0) {
                packageInfo.signatures[0] = _signature
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                if (packageInfo.signingInfo != null) {
                    val signaturesArray = packageInfo.signingInfo.apkContentsSigners
                    signaturesArray[0] = _signature
                }
            }
            return packageInfo
        }

        fun findField(_clazz: Class<*>, _fieldName: String): Field {
            var clazz = _clazz
            try {
                val field: Field = clazz.getDeclaredField(_fieldName)
                field.isAccessible = true
                return field
            } catch (e: NoSuchFieldException) {
                while (true) {
                    clazz = clazz.superclass
                    if (clazz == null || clazz.equals(Object::class.java)) {
                        break
                    }
                    try {
                        val field: Field = clazz.getDeclaredField(_fieldName)
                        field.isAccessible = true
                        return field
                    } catch (e: NoSuchFieldException) {

                    }
                }
                throw e
            }
        }

        fun addHiddenApiExemptions() {
            val forName = Class::class.java.getDeclaredMethod("forName", String::class.java)
            val getDeclaredMethod = Class::class.java.getDeclaredMethod(
                "getDeclaredMethod",
                String::class.java,
                arrayOf<Class<*>>()::class.java
            )

            val vmRuntimeClass = forName.invoke(null, "dalvik.system.VMRuntime") as Class<*>
            val getRuntime = getDeclaredMethod.invoke(vmRuntimeClass, "getRuntime", null) as Method
            val setHiddenApiExemptions = getDeclaredMethod.invoke(
                vmRuntimeClass,
                "setHiddenApiExemptions",
                arrayOf(arrayOf<String>()::class.java)
            ) as Method

            val vmRuntime = getRuntime.invoke(null)

            setHiddenApiExemptions.invoke(vmRuntime, arrayOf("L"))
        }
    }
}